### PR TITLE
Improve documentation and examples for `SchemaAdapterFactory`, make `record_batch` "hygenic"

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A

## Rationale for this change

When discussing schema adapting with @ccciudatu's on https://github.com/datafusion-contrib/datafusion-table-providers/pull/147#issuecomment-2424745760 I realized there were several levels of indirection that made it non trivial to figure out to use these classes to adapt schemas and how the different factories/methods are related to each other

## What changes are included in this PR?

1. Add examples
2. Add docs

## Are these changes tested?

CI

### CODE REVIEW

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 1]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
     ) -> datafusion_common::Result<RecordBatch>;
 }
 
-/// Implementation of [`SchemaAdapterFactory`] that maps columns by name
-/// and casts columns to the expected type.
+/// Default  [`SchemaAdapterFactory`] for mapping schemas.
```
Comment (alamb) [inline datafusion/core/src/datasource/schema_adapter.rs:110]: Screenshot 2024-10-22 at 10 55 58 AM

**Comment by alamb** [comment]
FYI @itsjunetime as you also previously worked with this code

File: datafusion/common/src/test_util.rs [hunk 2]
```diff
@@ -347,7 +347,7 @@ macro_rules! record_batch {
             let batch = arrow_array::RecordBatch::try_new(
                 schema,
                 vec![$(
-                    create_array!($type, $values),
+                    $crate::create_array!($type, $values),
```
Comment (alamb) [inline datafusion/common/src/test_util.rs:350]: Without this the macro errors if `create_array` is not also `use`d

File: datafusion/core/tests/macro_hygiene/mod.rs [hunk 3]
```diff
@@ -37,3 +37,13 @@ mod plan_datafusion_err {
...
     }
 }
+
+mod record_batch {
+    // NO other imports!
+    use datafusion_common::record_batch;
+
+    #[test]
+    fn test_macro() {
+        record_batch!(("column_name", Int32, vec![1, 2, 3])).unwrap();
```
Comment (alamb) [inline datafusion/core/tests/macro_hygiene/mod.rs:47]: test that we can use `record_batch` without any other imports

fyi @timsaucer

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 4]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
...
+///
+/// This can be used to adapt file-level record batches to a table schema and
+/// implement schema evolution.
+///
+/// Given an input file schema and a table schema, this factor can make
+/// [`SchemaMapper`]s that:
+///
+/// 1. Reorder columns
+/// 2. Cast columns to the correct type
+/// 3. Fill missing columns with nulls
```
Comment (crepererum) [inline datafusion/core/src/datasource/schema_adapter.rs:120]: What happens if the field is not nullable? I've looked through the code but couldn't find an explicit check for this, so does this error or does this produce invalid record batches? :thinking:

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 5]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
...
+/// # use datafusion::datasource::schema_adapter::{DefaultSchemaAdapterFactory, SchemaAdapterFactory};
+/// # use datafusion_common::record_batch;
+/// // Table has fields "a" and "b"
+/// let table_schema = Schema::new(vec![
+///     Field::new("a", DataType::Int32, true),
+///     Field::new("b", DataType::Utf8, true),
+///     Field::new("c", DataType::Utf8, true),
+/// ]);
+///
+/// // The file provides only fields "b" and "c" that oder
```
Comment (itsjunetime) [inline datafusion/core/src/datasource/schema_adapter.rs:156]:
```suggestion
/// // The file provides only fields "b" and "c" that order
```

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 6]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
...
+/// let projected_table_schema = table_schema.project(&[2, 1]).unwrap();
```
Comment (itsjunetime) [inline datafusion/core/src/datasource/schema_adapter.rs:157]: This looks like it's projecting fields `c` and `b`, in that order. Is the above comment wrong about which order they're being projected or am I misunderstanding what this is doing?

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 7]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
...
+///
+/// // create an adapter for the table schema and file schema
+/// let adapter = DefaultSchemaAdapterFactory.create(
+///   Arc::new(table_schema),
+///   Arc::new(projected_table_schema)
```
Comment (itsjunetime) [inline datafusion/core/src/datasource/schema_adapter.rs:162]:
```suggestion
///   Arc::new(projected_table_schema),
///   Arc::new(table_schema)
```

**Review by itsjunetime — COMMENTED** [review]
This looks generally good, but I'm a bit confused about some of the specifics of the doc test. I feel like it shouldn't work as written (due to the fields and projected schema vs normal schema being switched around), unless I'm missing something ofc, which makes me also wonder if doc-tests aren't being run as part of the test portion of CI. Am I confused about the doc test being wrong? Or should we set up doc tests to run with CI?

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 8]
```diff
@@ -99,8 +107,86 @@ pub trait SchemaMapper: Debug + Send + Sync {
...
+/// ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+///  ┌───────┐   ┌───────┐ │                  ┌───────┐   ┌───────┐   ┌───────┐ │
+/// ││  1.0  │   │ "foo" │                   ││ NULL  │   │ "foo" │   │ "1.0" │
+///  ├───────┤   ├───────┤ │ Schema mapping   ├───────┤   ├───────┤   ├───────┤ │
+/// ││  2.0  │   │ "bar" │                   ││  NULL │   │ "bar" │   │ "2.0" │
+///  └───────┘   └───────┘ │────────────────▶ └───────┘   └───────┘   └───────┘ │
+/// │                                        │
+///  column "c"  column "b"│                  column "a"  column "b"  column "c"│
+/// │ Float64       Utf8                     │  Int32        Utf8        Utf8
+///  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs:134]: I use https://monodraw.helftone.com/ (though I hear https://asciiflow.com is pretty good too)

I feel like I should get a commission for all the times I recommend Monodraw 😆

File: datafusion/core/src/datasource/schema_adapter.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs:120]: It errors -- I will add a test / improve the docs

**Comment by alamb** [comment]
> Am I confused about the doc test being wrong? Or should we set up doc tests to run with CI?

The doctest passes for me locally

```shell
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$ cargo test --doc -p datafusion -- schema_adapter
...
running 1 test
test datafusion/core/src/datasource/schema_adapter.rs - datasource::schema_adapter::DefaultSchemaAdapterFactory (line 144) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 97 filtered out; finished in 1.23s
```

I am pretty sure that the CI runs the doc tests as well. Here is this newly added test running in this PR:
https://github.com/apache/datafusion/actions/runs/11463343177/job/31897653058?pr=13063#step:5:921

However, your catch on the doc test is excellent and I will fix

File: datafusion/core/src/datasource/schema_adapter.rs (hunk 7 above)
Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs:162]: wow -- you are totally right. 🤦

Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs:162]: I don't really know how the previous version worked, but basically this stemmed from my misunderstanding of how this code should be used. I have updated now with more documentation

File: datafusion/core/src/datasource/schema_adapter.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs:126]: In [b065efe](https://github.com/apache/datafusion/pull/13063/commits/b065efe0456ed88bb7c89aff293acd71eef9cef4)

File: datafusion/core/src/datasource/schema_adapter.rs [hunk 9]
```diff
@@ -79,11 +87,17 @@ pub trait SchemaAdapter: Send + Sync {
...
 pub trait SchemaMapper: Debug + Send + Sync {
-    /// Adapts a `RecordBatch` to match the `table_schema` using the stored
-    /// mapping and conversions.
+    /// Adapts a `RecordBatch` to match the `table_schema`
+    ///
+    /// # Errors:
+    ///
+    /// * If a column in the table schema is non-nullable but is not present
+    ///   in the file schema (i.e. it is missing), this method tries to fill it
+    ///   with nulls resulting in a schema error
```
Comment (crepererum) [inline datafusion/core/src/datasource/schema_adapter.rs:100]: I think this is specific to the `DefaultSchemaAdapterFactory` and should be moved to there. You could imagine that someone writes a mapper that fills in missing columns with placeholder values or intra-/extrapolates the data.

Comment (alamb) [reply datafusion/core/src/datasource/schema_adapter.rs]: It is a good call -- tried to clarify in d966fd112

**Review by crepererum — APPROVED** [review]
needs a clippy fix, but otherwise OK :+1:

**Pull request merged**